File: uo/player.rb
#
#  GemUO
#  $Id$
#
#  (c) 2005 Max Kellermann <max@duempel.org>
#
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; version 2 of the License.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#

require 'uo/skills'

module GemUO
    LOCK_UP = 0
    LOCK_DOWN = 1
    LOCK_LOCKED = 2
    LOCK_NAMES = [ "Up", "Down", "Locked" ]

    STAT_STRENGTH = 0
    STAT_DEXTERITY = 1
    STAT_INTELLIGENCE = 2
    STAT_NAMES = ['Strength', 'Dexterity', 'Intelligence']

    SERIAL_PLAYER = 0x80000000

    class SkillValue
        def initialize(id, value, base, lock, cap)
            @id = id
            @value = value
            @base = base
            @lock = lock
            @cap = cap
        end

        def id
            @id
        end
        def name
            SKILL_NAMES[@id] || @id
        end
        def value
            @value
        end
        def base
            @base
        end
        def lock
            @lock
        end
        def cap
            @cap
        end

        def <=>(other)
            raise TypeError.new unless other.instance_of?(SkillValue)
            return @base <=> other.base
        end
        def to_s
            "[id=#{@id} value=#{@value} base=#{@base} lock=#{@lock} cap=#{@cap}]"
        end
    end

    class Walk
        def initialize(mobile)
            @mobile = mobile
            @next_seq = 0
        end
        def walk(direction)
            return unless @mobile.position
            return if @seq != nil
            @direction = direction & 0x7
            @seq = @next_seq
            @next_seq += 1
            @next_seq = 1 if @next_seq >= 0x100
            return GemUO::Packet::Walk.new(@direction, @seq)
        end
        def walk_reject(seq, x, y, z, direction)
            return unless @mobile.position

            # XXX resync when seq mismatch?
            @seq = nil
            @direction = nil
            @next_seq = 0
            @mobile.position = Position.new(x, y, z, direction)
        end
        def walk_ack(seq, notoriety)
            return unless @mobile.position

            if @seq != seq
                # XXX resync when seq mismatch?
            end

            oldpos = @mobile.position
            if oldpos.direction == @direction
                x, y = oldpos.x, oldpos.y
                case @direction
                when NORTH
                    y -= 1

                when NORTH_EAST
                    x += 1
                    y -= 1

                when EAST
                    x += 1

                when SOUTH_EAST
                    x += 1
                    y += 1

                when SOUTH
                    y += 1

                when SOUTH_WEST
                    x -= 1
                    y += 1

                when WEST
                    x -= 1

                when NORTH_WEST
                    x -= 1
                    y -= 1
                end

                @mobile.position = Position.new(x, y, oldpos.z, @direction)
            else
                @mobile.position = Position.new(oldpos.x, oldpos.y, oldpos.z,
                                                @direction)
            end

            @seq = nil
            @direction = nil
        end
    end
end

File: src/gemuo/player.py
#
#  GemUO
#
#  (c) 2005-2010 Max Kellermann <max@duempel.org>
#
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; version 2 of the License.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#

from uo.skills import SKILL_NAMES
from uo.entity import *
import uo.packets as p
from gemuo.entity import Position

class SkillValue:
    def __init__(self, id, value, base, lock, cap):
        self.id = id
        self.value = value
        self.base = base
        self.lock = lock
        self.cap = cap

    def name():
        if self.id in SKILL_NAMES:
            return SKILL_NAMES[self.id]
        else:
            return str(self.id)

class WalkQueue:
    def __init__(self):
        self.clear()

    def length(self):
        return len(self._queue)

    def _next_seq(self):
        seq = self._seq
        self._seq += 1
        if self._seq >= 0x100:
            self._seq = 1
        return seq

    def request(self):
        if len(self._queue) >= 3:
            return None

        seq = self._next_seq()
        self._queue.append(seq)
        return seq

    def ack(self, seq):
        while len(self._queue) > 0:
            i, self._queue = self._queue[0], self._queue[1:]
            if i == seq:
                return True
        return False

    def clear(self):
        self._queue = []
        self._seq = 0

    def reject(self, seq):
        self.clear()

class Walk:
    def __init__(self, mobile):
        self._mobile = mobile
        self._queue = WalkQueue()

    def finished(self):
        return self._queue.length() == 0

    def walk(self, direction):
        assert isinstance(direction, int)

        seq = self._queue.request()
        if seq is None: return None

        self._move_player(direction & 0x7)
        return p.WalkRequest(direction, seq)

    def walk_reject(self, seq, x, y, z, direction):
        self._queue.reject(seq)
        self._mobile.position = Position(x, y, z, direction)

    def walk_ack(self, seq, notoriety):
        if not self._queue.ack(seq):
            # XXX resync when seq mismatch?
            print "WalkAck out of sync"

    def _move_player(self, direction):
        oldpos = self._mobile.position
        if oldpos.direction == direction:
            x, y = oldpos.x, oldpos.y
            if direction == NORTH:
                y -= 1
            elif direction == NORTH_EAST:
                x += 1
                y -= 1
            elif direction == EAST:
                x += 1
            elif direction == SOUTH_EAST:
                x += 1
                y += 1
            elif direction == SOUTH:
                y += 1
            elif direction == SOUTH_WEST:
                x -= 1
                y += 1
            elif direction == WEST:
                x -= 1
            elif direction == NORTH_WEST:
                x -= 1
                y -= 1

            self._mobile.position = Position(x, y, oldpos.z, direction)
        else:
            self._mobile.position = Position(oldpos.x, oldpos.y, oldpos.z,
                                             direction)

    def move_player(self, direction):
        self._queue.clear()
        self._move_player(direction)
